### PR TITLE
Fix the inconsistency between local delete and remote clear

### PR DESCRIPTION
**Pull request by clarenceli-msft**
[AB#4612](https://dev.azure.com/fluidframework/internal/_workitems/edit/4612)

When encountering a remote clear operation, `clearExceptPendingKeys` will be called to retain the local modified but unack'd keys, however, it just treats all unack'd operations as `set`. For instance, if a local delete `delete("key1")` coincides with the remote clear, it will clear the map first, then force a `set("key1", undefined)`. It is necessary to determine if the unack'd operation is `delete` or `set` operation, this can be achieved by checking if the affected key still exists in the data or not.

### CODE REVIEW

File: packages/dds/map/src/mapKernel.ts [hunk 1]
```diff
@@ -547,8 +547,10 @@ export class MapKernel {
 		// we will get the value for the pendingKeys and clear the map
 		const temp = new Map<string, ILocalValue>();
 		for (const key of this.pendingKeys.keys()) {
-			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-			temp.set(key, this.data.get(key)!);
+			// Verify if the pending operation is a delete op, no need to retain it if so
```
Comment (Abe27342) [inline packages/dds/map/src/mapKernel.ts:550]:
```suggestion
			// Verify if the most recent pending operation is a delete op, no need to retain it if so.
			// This ensures the map size remains consistent.
```

**Review by Abe27342 — APPROVED** [review]
I don't think this fully works: if the map user calls `.set("foo", undefined)` rather than `.delete("foo")`, we won't be able to distinguish those cases based on the pending keys check you've added, but they have different effects on the map size. Could you file some work to potentially resolve that? I don't think it's particularly high-priority, since it's kinda edge-casey.

This fix looks good otherwise. Approving despite the above issue as it should make map's behavior better for the majority of cases.

**Comment by clarenceli-msft** [comment]
> I don't think this fully works: if the map user calls `.set("foo", undefined)` rather than `.delete("foo")`, we won't be able to distinguish those cases based on the pending keys check you've added, but they have different effects on the map size. Could you file some work to potentially resolve that? I don't think it's particularly high-priority, since it's kinda edge-casey.
> 
> This fix looks good otherwise. Approving despite the above issue as it should make map's behavior better for the majority of cases.

I think these two cases can be distinguished by the current logic, such as `this.data.has(key)` will return true if the operation is `set('foo', undefined)` because the key `foo` is actually inserted into the data, conversely it will return false if the operation is 'delete('foo')'. And yes it is an important corner case, have added it to the test suites

**Comment by Abe27342** [comment]
> t logic, such as `this.data.has(key)` will return true if the operation is `set('foo', undefined)` because the key `foo` is actually i

ah yeah, of course. I wouldn't be surprised if we had other bugs of a similar nature due to similar inconsistencies. in any case, :shipit: to this change.

**Pull request merged**